# 1. Introduction

Assume I own a consulting firm that provides location services for clients. The clients can be anyone who believe location matters for their business. They can be restaurants, banks, fitness centers, barbarshops and other shops.

During the Pandemic, a lot of restaurants were shut down, and some of them even had to close permanently. As business shutdown order is lifted, new restaurants are emerging to meet people’s demand for food in big cities. In this capstone project, I take an Italian restaurant as example of problem to solve. Now one of my clients is going to open an Italian restaurant in Boston, and he wants us to help him find the best location of three candidate locations.

## Audiance

Because the business we can serve can be all sorts of business that have demand of best location, our audiance is companies, organizations or business persons who want to find the best location to run their business with high performance.

## Problem to solve

The problem to solve is to look for the best location for an Italian restaurant. 

## Criteria of the best location

There are different criteria for different business that define what is the best location. In this project, the criteria of the best location for the Italian restaurant are as follows.

1. There are no more than two Italian Restaurants within 800 meters.
2. There is at least one big park within 800 meters. 
3. There is at least one big mall within 800 meters.


# 2. Data

1. Assume we already know the addresses of three candidate locations in Boston (they could be available rental places in the same time for restaurant business)
2. Park and restaurant data will be obtained from Foursquare that provides venues near the three candidate locations.
3. Source of Boston zipcode data: https://bostonopendata-boston.opendata.arcgis.com/datasets/53ea466a189b4f43b3dfb7b38fa7f3b6_1
4. Source of Massachusetts zipcode and latitude/longitude: https://public.opendatasoft.com/explore/dataset/us-zip-code-latitude-and-longitude/table/?q=MA
5. Shopping center data will be obtained from this webpage https://www.bostoncentral.com/shopping.php.


# 3. Methodology

In [183]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import csv
import geocoder

# Explore Boston by zip code

## Prepare data : (1) Read Boston zip code information from a csv file

In [185]:
df_Boston_zipcode = pd.read_csv('Boston/Boston_Zip_Codes.csv')
#df_Boston_zipcode.info
df_Boston_zipcode.head()

OBJECTID  ZIP5   ShapeSTArea  ShapeSTLength
0         1  2134  3.721936e+07    40794.18240
1         2  2125  6.476052e+07    62224.52144
2         3  2110  6.637284e+06    18358.21350
3         4  2118  3.116158e+07    32353.40762
4         5  2126  6.078585e+07    45488.39471

In [186]:
# Rename column ZIP5 to ZIP

df_Boston_zipcode.rename(columns={'ZIP5':'Zip'}, inplace=True)
df_Boston_zipcode.head()

OBJECTID   Zip   ShapeSTArea  ShapeSTLength
0         1  2134  3.721936e+07    40794.18240
1         2  2125  6.476052e+07    62224.52144
2         3  2110  6.637284e+06    18358.21350
3         4  2118  3.116158e+07    32353.40762
4         5  2126  6.078585e+07    45488.39471

In [187]:
df_Boston_zipcode.shape

(43, 4)

## Prepare data : (2) Read zip code and latitude/longitude of Massachusetts from a csv file. It includes Boston lat/lng information.

In [19]:
df_MA_zipcode_latlng = pd.read_csv('Boston/MA_Zip_Codes_latlng.csv',sep=';')
#df_Boston_zipcode.info
print(df_MA_zipcode_latlng.head())

    Zip            City State   Latitude  Longitude  Timezone  \
0  1085       Westfield    MA  42.133642  -72.75029        -5   
1  1340         Colrain    MA  42.673371  -72.73104        -5   
2  1886        Westford    MA  42.592086  -71.43754        -5   
3  2358  North Pembroke    MA  41.805219  -70.62642        -5   
4  1754         Maynard    MA  42.430781  -71.45594        -5   

   Daylight savings time flag             geopoint  
0                           1  42.133642,-72.75029  
1                           1  42.673371,-72.73104  
2                           1  42.592086,-71.43754  
3                           1  41.805219,-70.62642  
4                           1  42.430781,-71.45594  


## Prepare data : (3) Join Boston zip codes with MA zip code & lat/lng information

In [194]:
df_Boston_zip_latlng = pd.merge(df_Boston_zipcode, df_MA_zipcode_latlng, on='Zip', how='left')

In [195]:
df_Boston_zip_latlng.head()


OBJECTID   Zip   ShapeSTArea  ShapeSTLength      City State   Latitude  \
0         1  2134  3.721936e+07    40794.18240   Allston    MA  42.355147   
1         2  2125  6.476052e+07    62224.52144    Boston    MA  42.316852   
2         3  2110  6.637284e+06    18358.21350    Boston    MA  42.356532   
3         4  2118  3.116158e+07    32353.40762    Boston    MA  42.338724   
4         5  2126  6.078585e+07    45488.39471  Mattapan    MA  42.272098   

   Longitude  Timezone  Daylight savings time flag             geopoint  
0  -71.13164        -5                           1  42.355147,-71.13164  
1  -71.05811        -5                           1  42.316852,-71.05811  
2  -71.05365        -5                           1  42.356532,-71.05365  
3  -71.07276        -5                           1  42.338724,-71.07276  
4  -71.09426        -5                           1  42.272098,-71.09426

In [196]:
df_Boston_zip_latlng.shape

(43, 11)

## Prepare data: (4) Create dataframe of Boston with zipcode and latlng

In [197]:
Boston_zip_latlng = df_Boston_zip_latlng[['Zip', 'Latitude', 'Longitude']]
Boston_zip_latlng.head()

Zip   Latitude  Longitude
0  2134  42.355147  -71.13164
1  2125  42.316852  -71.05811
2  2110  42.356532  -71.05365
3  2118  42.338724  -71.07276
4  2126  42.272098  -71.09426

In [198]:
Boston_zip_latlng.shape

(43, 3)

In [199]:
Boston_zip_latlng.groupby('Zip').count()

Latitude  Longitude
Zip                      
2021         1          1
2026         1          1
2108         1          1
2109         1          1
2110         1          1
2111         1          1
2113         1          1
2114         1          1
2115         1          1
2116         1          1
2118         1          1
2119         1          1
2120         1          1
2121         1          1
2122         1          1
2124         1          1
2125         1          1
2126         1          1
2127         1          1
2128         1          1
2129         1          1
2130         1          1
2131         1          1
2132         1          1
2134         1          1
2135         1          1
2136         1          1
2151         1          1
2152         1          1
2163         1          1
2186         1          1
2199         1          1
2203         1          1
2210         1          1
2215         1          1
2459         1          1
2467         7          7

## Find duplicate records at zip code 2467

## Prepare data : (5) Remove duplicate records at zip code 2467

In [200]:
Boston_zip_latlng.drop_duplicates(inplace=True)

c:\users\hmzhu\files(x86)\py_virtual_env\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [201]:
Boston_zip_latlng.shape

(37, 3)

## Create a map of Boston zipcodes

In [203]:
# Get latitude and longitude of Boston certroid in order to center map

from geopy.geocoders import Nominatim

address = 'Boston, Massachusetts, US'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Boston are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Boston are 42.3602534, -71.0582912.


In [204]:
import folium # map rendering library

Boston_map = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers and popup wondin to map
for zipcode, lat, lng in zip(Boston_zip_latlng['Zip'], Boston_zip_latlng['Latitude'], Boston_zip_latlng['Longitude']):
    label = '{}'.format(zipcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(Boston_map)  
    
Boston_map

## In order to determine the best location for an Italian restaurant, we need to better understand the geographic distribution of all sorts of businesses in Boston. To gain the knowledge of what kinds of businesses are in each zipcode in Boston, we need to obtain venue information in each zipcode.

# Request venues through Foursquare

In [206]:
CLIENT_ID = 'S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J' # your Foursquare ID
CLIENT_SECRET = 'OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

## Define a function to return 100 venues witin 500 meters from the center of each zipcode

In [227]:
def getNearbyVenues(zipcodes, latitudes, longitudes, radius=500):
    LIMIT = 100
    venues_list=[]
    for zipcode, lat, lng in zip(zipcodes, latitudes, longitudes):  
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        #print(url)

        results = requests.get(url).json()["response"]['groups'][0]['items']
        print (results[0])

        venues_list.append([(
            zipcode, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

## Obtain venues returned by Foursqure service

In [217]:
Boston_venues = getNearbyVenues(zipcodes=Boston_zip_latlng['Zip'],
                                   latitudes=Boston_zip_latlng['Latitude'],
                                   longitudes=Boston_zip_latlng['Longitude']
                                )
  

https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.355146999999995,-71.13164&radius=500&limit=100
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.316852000000004,-71.05811&radius=500&limit=100
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.356532,-71.05365&radius=500&limit=100
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.338724,-71.07276&radius=500&limit=100
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F

In [228]:
Boston_venues.head()

Zip   Latitude  Longitude                  Venue  Venue Latitude  \
0  2134  42.355147  -71.13164         Lulu's Allston       42.355068   
1  2134  42.355147  -71.13164        Kaju Tofu House       42.354329   
2  2134  42.355147  -71.13164  Fish Market Sushi Bar       42.353039   
3  2134  42.355147  -71.13164        BonChon Chicken       42.353105   
4  2134  42.355147  -71.13164        Mala Restaurant       42.352960   

   Venue Longitude           Venue Category  
0       -71.134107  Comfort Food Restaurant  
1       -71.132374        Korean Restaurant  
2       -71.132975         Sushi Restaurant  
3       -71.130921      Fried Chicken Joint  
4       -71.131033       Chinese Restaurant

In [225]:
Boston_venues.shape

(1534, 7)

In [220]:
Boston_venues['Venue Category'].unique()

array(['Comfort Food Restaurant', 'Korean Restaurant', 'Sushi Restaurant',
       'Fried Chicken Joint', 'Chinese Restaurant', 'Falafel Restaurant',
       'Diner', 'Indian Restaurant', 'Japanese Restaurant',
       'Thai Restaurant', 'Italian Restaurant', 'Thrift / Vintage Store',
       'Rock Club', 'Smoke Shop', 'Bakery', 'Restaurant',
       'Vietnamese Restaurant', 'Liquor Store', 'Hot Dog Joint',
       'Mexican Restaurant', 'Mediterranean Restaurant',
       'Asian Restaurant', 'Frozen Yogurt Shop', 'Pizza Place',
       'Tea Room', 'Vegetarian / Vegan Restaurant', 'Board Shop',
       'Dumpling Restaurant', 'Taco Place', 'Gastropub', 'Gym',
       'Electronics Store', 'Ice Cream Shop', 'Sandwich Place', 'Café',
       'Hookah Bar', 'Seafood Restaurant', 'Bubble Tea Shop',
       'Sports Bar', 'Pet Store', 'Dive Bar', 'Pharmacy',
       'Hardware Store', 'Shoe Store', 'Latin American Restaurant',
       'Donut Shop', 'Nightclub', 'Dance Studio', 'Fast Food Restaurant',
       'S

In [229]:
# one hot encoding
boston_onehot = pd.get_dummies(Boston_venues[['Venue Category']], prefix="", prefix_sep="")

# add zipcode column back to dataframe
boston_onehot['Zip'] = Boston_venues['Zip'] 

# move zipcode column to the first column
fixed_columns = [boston_onehot.columns[-1]] + list(boston_onehot.columns[:-1])
boston_onehot = boston_onehot[fixed_columns]

boston_onehot.head()

Zip  Accessories Store  African Restaurant  Airport  American Restaurant  \
0  2134                  0                   0        0                    0   
1  2134                  0                   0        0                    0   
2  2134                  0                   0        0                    0   
3  2134                  0                   0        0                    0   
4  2134                  0                   0        0                    0   

   Aquarium  Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0         0                 0            0           0                    0   
1         0                 0            0           0                    0   
2         0                 0            0           0                    0   
3         0                 0            0           0                    0   
4         0                 0            0           0                    0   

   ...  Trail  Udon Restaurant  Vegetarian / Vegan Restaurant  \
0  ...      0                0                              0   
1  ...      0                0                              0   
2  ...      0                0                              0   
3  ...      0                0                              0   
4  ...      0                0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  Winery  \
0                 0                      0         0          0       0   
1                 0                      0         0          0       0   
2                 0                      0         0          0       0   
3                 0                      0         0          0       0   
4                 0                      0         0          0       0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 230 columns]

In [230]:
boston_onehot.shape

(1534, 230)

In [231]:
# Normalize data in order to analyze
boston_grouped = boston_onehot.groupby('Zip').mean().reset_index()
boston_grouped

Zip  Accessories Store  African Restaurant   Airport  \
0   2021           0.000000            0.000000  0.000000   
1   2026           0.000000            0.000000  0.000000   
2   2108           0.000000            0.000000  0.000000   
3   2109           0.000000            0.000000  0.000000   
4   2110           0.000000            0.000000  0.000000   
5   2111           0.000000            0.000000  0.000000   
6   2113           0.000000            0.000000  0.000000   
7   2114           0.000000            0.000000  0.000000   
8   2115           0.000000            0.000000  0.000000   
9   2116           0.010000            0.000000  0.000000   
10  2118           0.000000            0.019608  0.000000   
11  2119           0.000000            0.166667  0.000000   
12  2120           0.000000            0.055556  0.000000   
13  2121           0.000000            0.000000  0.000000   
14  2122           0.000000            0.000000  0.000000   
15  2124           0.000000            0.000000  0.000000   
16  2125           0.000000            0.000000  0.000000   
17  2126           0.000000            0.000000  0.000000   
18  2127           0.000000            0.000000  0.000000   
19  2128           0.000000            0.000000  0.047619   
20  2129           0.000000            0.000000  0.000000   
21  2130           0.045455            0.000000  0.000000   
22  2131           0.000000            0.000000  0.000000   
23  2132           0.000000            0.000000  0.000000   
24  2134           0.000000            0.000000  0.000000   
25  2135           0.000000            0.000000  0.000000   
26  2136           0.000000            0.000000  0.000000   
27  2151           0.000000            0.000000  0.000000   
28  2152           0.000000            0.000000  0.000000   
29  2163           0.000000            0.000000  0.000000   
30  2186           0.000000            0.000000  0.000000   
31  2199           0.010000            0.000000  0.000000   
32  2203           0.000000            0.000000  0.000000   
33  2210           0.000000            0.000000  0.000000   
34  2215           0.000000            0.000000  0.000000   
35  2459           0.000000            0.000000  0.000000   
36  2467           0.000000            0.000000  0.000000   

    American Restaurant  Aquarium  Arepa Restaurant  Art Gallery  Art Museum  \
0              0.000000  0.000000          0.000000     0.000000    0.000000   
1              0.000000  0.000000          0.000000     0.000000    0.000000   
2              0.034091  0.000000          0.000000     0.000000    0.000000   
3              0.020000  0.000000          0.000000     0.000000    0.000000   
4              0.022989  0.034483          0.000000     0.000000    0.011494   
5              0.020000  0.000000          0.000000     0.000000    0.000000   
6              0.010000  0.000000          0.000000     0.000000    0.000000   
7              0.050000  0.000000          0.000000     0.000000    0.000000   
8              0.030303  0.000000          0.000000     0.030303    0.121212   
9              0.020000  0.000000          0.000000     0.000000    0.000000   
10             0.019608  0.000000          0.019608     0.000000    0.000000   
11             0.333333  0.000000          0.000000     0.000000    0.000000   
12             0.000000  0.000000          0.000000     0.055556    0.000000   
13             0.000000  0.000000          0.000000     0.000000    0.000000   
14             0.000000  0.000000          0.000000     0.000000    0.000000   
15             0.000000  0.000000          0.000000     0.000000    0.000000   
16             0.000000  0.000000          0.000000     0.000000    0.000000   
17             0.000000  0.000000          0.000000     0.000000    0.000000   
18             0.034483  0.000000          0.000000     0.000000    0.000000   
19             0.000000  0.000000          0.000000     0.000000    0.000000   
20    

## Explore what businesses are the most common in a zipcode area

In [233]:
# Check top 5 venue categories and their scores in each zipcode

num_top_venues = 5

for zipcode in boston_grouped['Zip']:
    print("----"+str(zipcode)+"----")
    temp = boston_grouped[boston_grouped['Zip'] == zipcode].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----2021----
                venue  freq
0  Athletics & Sports  0.50
1     Paintball Field  0.25
2                Farm  0.25
3   Accessories Store  0.00
4   Other Repair Shop  0.00


----2026----
            venue  freq
0  Farmers Market   0.2
1           Diner   0.2
2             Gym   0.2
3           Track   0.2
4     High School   0.2


----2108----
                 venue  freq
0          Pizza Place  0.05
1          Coffee Shop  0.05
2        Historic Site  0.03
3           Steakhouse  0.03
4  American Restaurant  0.03


----2109----
                venue  freq
0  Italian Restaurant  0.22
1  Seafood Restaurant  0.10
2              Bakery  0.06
3                Park  0.05
4       Historic Site  0.04


----2110----
                venue  freq
0       Boat or Ferry  0.11
1  Seafood Restaurant  0.07
2                Park  0.06
3               Hotel  0.05
4       Historic Site  0.05


----2111----
                venue  freq
0  Chinese Restaurant  0.13
1              Bakery  0.08
2    A

## Define a function return_most_common_venues() to return the most common businesses in each zipcode

In [234]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

## Obtain top 10 most common venues

In [236]:
num_top_venues = 10
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zipcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
zipcode_10_most = pd.DataFrame(columns=columns)
zipcode_10_most['Zipcode'] = boston_grouped['Zip']

for ind in np.arange(boston_grouped.shape[0]):
    zipcode_10_most.iloc[ind, 1:] = return_most_common_venues(boston_grouped.iloc[ind, :], num_top_venues)

zipcode_10_most.head()

Zipcode 1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     2021    Athletics & Sports       Paintball Field                  Farm   
1     2026                 Diner                 Track        Farmers Market   
2     2108           Coffee Shop           Pizza Place    Italian Restaurant   
3     2109    Italian Restaurant    Seafood Restaurant                Bakery   
4     2110         Boat or Ferry    Seafood Restaurant                  Park   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0           Yoga Studio    Falafel Restaurant            Food Truck   
1                   Gym           High School           Yoga Studio   
2        Sandwich Place            Restaurant                 Plaza   
3                  Park         Historic Site           Pizza Place   
4         Historic Site                 Hotel       Harbor / Marina   

     7th Most Common Venue       8th Most Common Venue 9th Most Common Venue  \
0             Food Service                  Food Court     Food & Drink Shop   
1       Falafel Restaurant                Food Service            Food Court   
2  New American Restaurant                  Steakhouse   American Restaurant   
3                     Café  Tourist Information Center         Grocery Store   
4              Coffee Shop            Asian Restaurant    Italian Restaurant   

  10th Most Common Venue  
0                   Food  
1      Food & Drink Shop  
2          Historic Site  
3                  Hotel  
4               Aquarium

## Create clusters with k-mean with k=5

In [238]:
# import k-means
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

boston_grouped_clustering = boston_grouped.drop('Zip', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(boston_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 2, 1, 1, 1, 1, 1, 1, 1, 1])

In [239]:
Boston_zip_latlng.columns

Index(['Zip', 'Latitude', 'Longitude'], dtype='object')

In [240]:
Boston_zip_latlng.shape

(37, 3)

In [241]:
# add clustering labels
zipcode_10_most.insert(0, 'Cluster Labels', kmeans.labels_)

In [242]:
zipcode_10_most.head()

Cluster Labels  Zipcode 1st Most Common Venue 2nd Most Common Venue  \
0               4     2021    Athletics & Sports       Paintball Field   
1               2     2026                 Diner                 Track   
2               1     2108           Coffee Shop           Pizza Place   
3               1     2109    Italian Restaurant    Seafood Restaurant   
4               1     2110         Boat or Ferry    Seafood Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Farm           Yoga Studio    Falafel Restaurant   
1        Farmers Market                   Gym           High School   
2    Italian Restaurant        Sandwich Place            Restaurant   
3                Bakery                  Park         Historic Site   
4                  Park         Historic Site                 Hotel   

  6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
0            Food Truck             Food Service                  Food Court   
1           Yoga Studio       Falafel Restaurant                Food Service   
2                 Plaza  New American Restaurant                  Steakhouse   
3           Pizza Place                     Café  Tourist Information Center   
4       Harbor / Marina              Coffee Shop            Asian Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0     Food & Drink Shop                   Food  
1            Food Court      Food & Drink Shop  
2   American Restaurant          Historic Site  
3         Grocery Store                  Hotel  
4    Italian Restaurant               Aquarium

In [243]:
zipcode_10_most.rename(columns={"Zipcode":"Zip"}, inplace=True)

In [244]:
zipcode_10_most.head()

Cluster Labels   Zip 1st Most Common Venue 2nd Most Common Venue  \
0               4  2021    Athletics & Sports       Paintball Field   
1               2  2026                 Diner                 Track   
2               1  2108           Coffee Shop           Pizza Place   
3               1  2109    Italian Restaurant    Seafood Restaurant   
4               1  2110         Boat or Ferry    Seafood Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Farm           Yoga Studio    Falafel Restaurant   
1        Farmers Market                   Gym           High School   
2    Italian Restaurant        Sandwich Place            Restaurant   
3                Bakery                  Park         Historic Site   
4                  Park         Historic Site                 Hotel   

  6th Most Common Venue    7th Most Common Venue       8th Most Common Venue  \
0            Food Truck             Food Service                  Food Court   
1           Yoga Studio       Falafel Restaurant                Food Service   
2                 Plaza  New American Restaurant                  Steakhouse   
3           Pizza Place                     Café  Tourist Information Center   
4       Harbor / Marina              Coffee Shop            Asian Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0     Food & Drink Shop                   Food  
1            Food Court      Food & Drink Shop  
2   American Restaurant          Historic Site  
3         Grocery Store                  Hotel  
4    Italian Restaurant               Aquarium

In [245]:
Boston_zip_latlng.head()

Zip   Latitude  Longitude
0  2134  42.355147  -71.13164
1  2125  42.316852  -71.05811
2  2110  42.356532  -71.05365
3  2118  42.338724  -71.07276
4  2126  42.272098  -71.09426

In [246]:
Boston_zip_latlng.head()

Zip   Latitude  Longitude
0  2134  42.355147  -71.13164
1  2125  42.316852  -71.05811
2  2110  42.356532  -71.05365
3  2118  42.338724  -71.07276
4  2126  42.272098  -71.09426

## In order to append location info to zicode_10_most, merge it with Boston_zip_latlng

In [247]:
# Merge zipcode_10_most with Boston_zip_latlng
boston_merged = Boston_zip_latlng
boston_merged = boston_merged.join(zipcode_10_most.set_index('Zip'), on='Zip')

boston_merged.head(37)

Zip   Latitude  Longitude  Cluster Labels  \
0   2134  42.355147 -71.131640               1   
1   2125  42.316852 -71.058110               1   
2   2110  42.356532 -71.053650               1   
3   2118  42.338724 -71.072760               1   
4   2126  42.272098 -71.094260               1   
5   2109  42.361477 -71.054170               1   
6   2021  42.166776 -71.134300               4   
7   2113  42.365028 -71.056360               1   
8   2130  42.309998 -71.111710               1   
9   2121  42.307448 -71.081270               1   
10  2119  42.324511 -71.084550               1   
11  2115  42.343080 -71.092680               1   
12  2163  42.367946 -71.120560               1   
13  2135  42.348418 -71.153490               1   
14  2199  42.347130 -71.082340               1   
15  2459  42.321197 -71.193000               0   
16  2124  42.287130 -71.071560               1   
17  2132  42.279432 -71.159800               1   
18  2114  42.361792 -71.067740               1   
19  2108  42.357903 -71.064080               1   
20  2136  42.252198 -71.125930               1   
21  2111  42.349838 -71.061010               1   
22  2210  42.347974 -71.044630               1   
23  2116  42.349622 -71.073720               1   
24  2131  42.284678 -71.130520               1   
25  2127  42.334990 -71.045620               1   
26  2120  42.332099 -71.096510               1   
27  2203  42.361485 -71.060364               1   
28  2215  42.346997 -71.102150               1   
29  2129  42.378619 -71.064950               1   
30  2128  42.378297 -71.025500               1   
31  2122  42.296298 -71.054540               1   
32  2151  42.416247 -71.005250               1   
33  2186  42.253074 -71.074960               3   
34  2026  42.244609 -71.165530               2   
35  2152  42.373047 -70.978030               1   
36  2467  42.321997 -71.173140               1   

              1st Most Common Venue 2nd Most Common Venue  \
0                 Korean Restaurant           Pizza Place   
1                               Bar                   Pub   
2                     Boat or Ferry    Seafood Restaurant   
3                       Pizza Place    Mexican Restaurant   
4                       Pizza Place          Home Service   
5                Italian Restaurant    Seafood Restaurant   
6                Athletics & Sports       Paintball Field   
7                Italian Restaurant           Pizza Place   
8                            Bakery           Coffee Shop   
9   Southern / Soul Food Restaurant           Supermarket   
10              American Restaurant                  Park   
11                      Pizza Place            Art Museum   
12             Gym / Fitness Center                  Park   
13                             Bank                Bakery   
14                   Clothing Store                 Hotel   
15                     Home Service            Print Shop   
16                    Deli / Bodega                  Park   
17                    Grocery Store           Pizza Place   
18                       Food Truck   American Restaurant   
19                      Coffee Shop           Pizza Place   
20                         Pharmacy        Discount Store   
21               Chinese Restaurant                Bakery   
22                       Restaurant                 Hotel   
23                              Spa  Gym / Fitness Center   
24              American Restaurant                Bakery   
25                      Coffee Shop            Sports Bar   
26                      Pizza Place            Donut Shop   
27               Italian Restaurant         Historic Site   
28               Mexican Restaurant   American Restaurant   
29                      Coffee Shop     Convenience Store   
30               Italian Restaurant  Brazilian Restaurant   
31                      Pizza Place   Arts & Crafts Store   
32               Italian Restaurant         Deli / Bodega   
33                            

### From the table above, we can conclude what types of venues each label denotes as follows

1. Label 0 - print shop, trail
2. Label 1 - all sorts of food and drink places
3. Label 2 - farm, gym, school
4. Label 3 - park, sport field
5. Label 4 - sport field, farm

## Create a cluster map based on cluster labels and location

In [293]:
# create cluster map

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

boston_map1 = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(boston_merged['Latitude'], boston_merged['Longitude'], boston_merged['Zip'], boston_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(boston_map1)
       
boston_map1

## Interprete cluster map

### The map tells us that there are all sorts of business in downtown areas in Boston except four zipcodes.  A lot of them are different cuisines of restaurants and drink shops, as well as shopping malls, schools etc. They are equally commercial areas. Therefore, there is no preference for candidate loactions at the city scale. 

### Assume we already have three candidate locations for a new Italian restaurant. Now we are going to check each of them at smaller scale, which is surrending area around them.

# Add three candidate locations to map

## First, obtain latitude & longitude of each candidate location

In [251]:
# Addresses of three candidate locations
locations = ['48 Deckard St, Boston, MA 02121', '52 Mt Vernon St, Boston, MA 02108', '270 W Fifth St, Boston, MA 02127']

# Get latitude and longitude for each candidate location
candidate_lat = []
candidate_lng = []

for i in range(len(locations)):
    print(i)
    address = locations[i]
    #geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    
    if (location is None):
        print ("None: " + address+ "\n")
    else:
        latitude = location.latitude
        longitude = location.longitude
        print(latitude)
        print(longitude)
        print("\n")
        candidate_lat.append(latitude)
        candidate_lng.append(longitude)
print(candidate_lat)
print(candidate_lng)


0
42.315778819228
-71.08507985113481


1
42.35810105
-71.0669259504036


2
42.3353948028169
-71.04899235211268


[42.315778819228, 42.35810105, 42.3353948028169]
[-71.08507985113481, -71.0669259504036, -71.04899235211268]


## Add candidate locations on map

In [294]:
for i in range(len(candidate_lat)):
    folium.Marker(
        #[lat, lon],
        #radius=6,
        location=[candidate_lat[i], candidate_lng[i]],
        #popup=locations[i],
        popup = 'Candidate location ' + str(i+1),
        icon=folium.Icon(color='orange',icon='utensils',prefix='fa')
        ).add_to(boston_map1)
    

In [296]:
boston_map1

## Find nearby venues of each candidate location within 800 meters

In [256]:
radius = 800
LIMIT = 200
candidate_venues = []

for i in range(len(candidate_lat)):
    # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            candidate_lat[i], 
            candidate_lng[i], 
            radius, 
            LIMIT)
        print(url)
        
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        candidate_venues.append([(
            i, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        
##
df_candidate_venues = pd.DataFrame([item for venue in candidate_venues for item in venue])

df_candidate_venues.columns = [
              'id', 
              'venue name', 
              'venue latitude', 
              'venue longitude', 
              'venue category']

df_candidate_venues.head()

https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.315778819228,-71.08507985113481&radius=800&limit=200
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.35810105,-71.0669259504036&radius=800&limit=200
https://api.foursquare.com/v2/venues/explore?&client_id=S3PRMDQX0MOCKBHZ40GXLRO3NWJ3EIOY0F0GIN52JDY5FQ0J&client_secret=OVMTVGTK5EW0WM5OIMT2CMJ4MGN5EQNUUKRKU1IRCCAQ0MCL&v=20180605&ll=42.3353948028169,-71.04899235211268&radius=800&limit=200


id                 venue name  venue latitude  venue longitude  \
0   0  Popeyes Louisiana Kitchen       42.318547       -71.082893   
1   0    The Merengue Restaurant       42.319199       -71.077655   
2   0               Roxbury YMCA       42.317791       -71.082789   
3   0                 Flames III       42.309020       -71.083061   
4   0                  Walgreens       42.316881       -71.082464   

         venue category  
0   Fried Chicken Joint  
1  Caribbean Restaurant  
2  Gym / Fitness Center  
3  Caribbean Restaurant  
4              Pharmacy

In [257]:
df_candidate_venues.shape

(183, 5)

In [258]:
df_candidate_venues.groupby('id').count()

venue name  venue latitude  venue longitude  venue category
id                                                             
0           23              23               23              23
1          100             100              100             100
2           60              60               60              60

## Select venues near candidate location 1 and put them into a dateframe

In [160]:
# select venues with id 0
candidate1_venues = df_candidate_venues[df_candidate_venues['id']==0]
candidate1_venues.head(50)

id                        venue name  venue latitude  venue longitude  \
0    0         Popeyes Louisiana Kitchen       42.318547       -71.082893   
1    0           The Merengue Restaurant       42.319199       -71.077655   
2    0                      Roxbury YMCA       42.317791       -71.082789   
3    0                        Flames III       42.309020       -71.083061   
4    0                         Walgreens       42.316881       -71.082464   
5    0                         Marshalls       42.319429       -71.081941   
6    0                      CVS pharmacy       42.309031       -71.081922   
7    0                             Shell       42.314149       -71.078780   
8    0               The Mall Of Roxbury       42.320220       -71.082122   
9    0                        Save-A-Lot       42.319490       -71.082307   
10   0                           Comcast       42.319016       -71.082436   
11   0         Brothers Super Liquors II       42.318698       -71.081536   
12   0               CROWN FRIED CHICKEN       42.319218       -71.081483   
13   0                  Prieto Empanadas       42.312618       -71.089684   
14   0            New Family Supermarket       42.312695       -71.089973   
15   0                           Dunkin'       42.309408       -71.082084   
16   0  Malcolm X Park Basketball Courts       42.320378       -71.089527   
17   0                       stop & Shop       42.309228       -71.081361   
18   0                        McDonald's       42.320352       -71.082228   
19   0                  Blue Hill Avenue       42.318970       -71.077568   
20   0                  Mattapans Finest       42.309326       -71.083130   
21   0               Melnea A Cass Arena       42.320970       -71.091059   
22   0         Melnea Cass Arena Complex       42.321377       -71.090576   

          venue category  
0    Fried Chicken Joint  
1   Caribbean Restaurant  
2   Gym / Fitness Center  
3   Caribbean Restaurant  
4               Pharmacy  
5       Department Store  
6               Pharmacy  
7            Gas Station  
8          Shopping Mall  
9          Grocery Store  
10     Electronics Store  
11          Liquor Store  
12   Fried Chicken Joint  
13   American Restaurant  
14     Convenience Store  
15            Donut Shop  
16      Basketball Court  
17           Supermarket  
18  Fast Food Restaurant  
19    Spanish Restaurant  
20        Cosmetics Shop  
21          Skating Rink  
22                   Gym

## Select venues near candidate location 2 and put them into a dateframe

In [264]:
# select venues near id 1
candidate2_venues = df_candidate_venues[df_candidate_venues['id']==1]
candidate2_venues.head()

id                          venue name  venue latitude  venue longitude  \
23   1                    Louisburg Square       42.358636       -71.068903   
24   1                 Tatte Bakery & Cafe       42.357904       -71.070439   
25   1                           Frog Pond       42.356134       -71.065672   
26   1                                Figs       42.357255       -71.070007   
27   1  Museum of African American History       42.360058       -71.065287   

    venue category  
23           Plaza  
24          Bakery  
25            Lake  
26     Pizza Place  
27  History Museum

## Select venues near candidate location 3 and put them into a dateframe

In [265]:
# select venues near id 2
candidate3_venues = df_candidate_venues[df_candidate_venues['id']==2]
candidate3_venues.head()

id                   venue name  venue latitude  venue longitude  \
123   2                         Capo       42.336070       -71.047079   
124   2   Loco Taqueria & Oyster Bar       42.336999       -71.047803   
125   2  Lincoln Tavern & Restaurant       42.336406       -71.047522   
126   2                   J.P. Licks       42.336932       -71.048436   
127   2                  Sweet Tooth       42.337901       -71.049432   

          venue category  
123   Italian Restaurant  
124   Mexican Restaurant  
125  American Restaurant  
126       Ice Cream Shop  
127         Dessert Shop

## Criteria of best Italian restaurant

1. There are no more than two Italian Restaurants within 800 meters.
2. There is at least one park within 800 meters. 
3. There is at least one mall within 800 meters.

## Define a function to check criteria for each location

In [270]:
def check_best_location(candidate_venues):
    i = 0  #number of Italian resraurant
    criterion1 = 'There is no Italian Restaurant within 800 meters'
    
    j = 0 # number of park
    criterion2 = 'There is no park within 800 meters'
    
    k = 0 # number of shopping mall
    criterion3 = 'There is no shopping mall within 800 meters'
    
    for venue in candidate_venues:
        
        #print(venue)
        if (venue.lower() == 'italian restaurant'):
            i = i+1
            criterion1 = str(i) + ' Italian Restaurant(s) found within 800 meters.'
            
        if(venue.lower() == 'park'):
            j = j+1
            criterion2 = str(j) + ' park(s) found within 800 meters.'
            
        if(venue.lower() == 'shopping mall'):
            k = k+1
            criterion3 = str(k) + ' shopping mall(s) found within 800 meters.'
            
    print(criterion1)
    print('\n')
    print(criterion2)
    print('\n')
    print(criterion3)
    print('\n')

# 4. Result

## (1) Results of candidate location 1

In [271]:
check_best_location(candidate1_category)

There is no Italian Restaurant within 800 meters


There is no park within 800 meters


1 shopping mall(s) found within 800 meters.




## (2) Results of candidate location 2

In [272]:
check_best_location(candidate2_category)

5 Italian Restaurant(s) found within 800 meters.


3 park(s) found within 800 meters.


There is no shopping mall within 800 meters




## (3) Results of candidate location 3

In [273]:
check_best_location(candidate3_category)

2 Italian Restaurant(s) found within 800 meters.


1 park(s) found within 800 meters.


There is no shopping mall within 800 meters




# 5. Discussion

### Result 1: There is one big shopping mall within 800 meters, and there is no Italian restaurant and park nearby.

### Result 2: There are five Italian restaurants within 800 meters. Even though there are three parks nearby, that still indicates potential fierce competition between these Italian restaurants. There is no shopping mall within 800 meters.

### Result 3: Near this candidate location, there are two Italian restaurants and one park within 800 meters, and there is no shopping mall either.

### It is not straightforward to make a final decision just based on the three results because they convey mixed messages. So further analysis is needed.  So I'm going to add shopping malls on the map, which will be helpful.

## Add shopping malls on map

### First, open Boston shopping mall data

In [274]:
# shopping malls - points
boston_shopping_mall = pd.read_csv('Boston/Boston_shopping_malls.csv')
#df_Boston_zipcode.info
boston_shopping_mall.head()

mall name             address     city  zipcode  lat  lng
0  Prudential Center Boston     800 Boylston St   Boston     2199  NaN  NaN
1              Copley Place  100 Huntington Ave   Boston     2116  NaN  NaN
2  Faneuil Hall Marketplace       4 S Market St   Boston     2109  NaN  NaN
3          South Bay Center       8 Allstate Rd   Boston     2118  NaN  NaN
4      Washington Park Mall   330 M.L.K.Jr Blvd   Boston     2119  NaN  NaN

### Fix incorrect addresses

In [276]:
#boston_shopping_mall.iloc(4) = {'address':'330 Martin Luther King Jr Blvd'}
boston_shopping_mall.at[4,"address"] = '330 Martin Luther King Jr Blvd'

In [277]:
boston_shopping_mall.at[9,"address"] = '24 Winter Place'
boston_shopping_mall.at[13,"address"] = '100 Cambridgeside Place'

In [279]:
boston_shopping_mall.head(16)

mall name                         address       city  \
0        Prudential Center Boston                 800 Boylston St     Boston   
1                    Copley Place              100 Huntington Ave     Boston   
2        Faneuil Hall Marketplace                   4 S Market St     Boston   
3                South Bay Center                   8 Allstate Rd     Boston   
4            Washington Park Mall  330 Martin Luther King Jr Blvd     Boston   
5               Longwood Galleria               400 Brookline Ave     Boston   
6                 Fresh Pond Mall                140 Clarendon St     Boston   
7      Memorial Convention Center                 900 Boylston St     Boston   
8        Stores At 500 Washington               500 Washington St     Boston   
9               Downtown Crossing                 24 Winter Place     Boston   
10                The Corner Mall               417 Washington St     Boston   
11                Mercantile Mall                      1 Cross St     Boston   
12                   Primax Store                    5 Blossom St     Boston   
13                 Cambridge Side         100 Cambridgeside Place  Cambridge   
14                     The Atrium                    50 Church St  Cambridge   
15  Porter Square Shopping Center                     37 White St  Cambridge   

    zipcode  lat  lng  
0      2199  NaN  NaN  
1      2116  NaN  NaN  
2      2109  NaN  NaN  
3      2118  NaN  NaN  
4      2119  NaN  NaN  
5      2215  NaN  NaN  
6      2116  NaN  NaN  
7      2199  NaN  NaN  
8      2111  NaN  NaN  
9      2108  NaN  NaN  
10     2129  NaN  NaN  
11     2109  NaN  NaN  
12     2114  NaN  NaN  
13     2141  NaN  NaN  
14     2138  NaN  NaN  
15     2140  NaN  NaN

### Obtain lat/lng of each mall through geocoders

In [9]:
from geopy.geocoders import Nominatim

address = 'Boston, Massachusetts, US'
geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Boston are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of Boston are 42.3602534, -71.0582912.


In [ ]:
# 42.3602534, -71.0582912.
# 42.3471718, -71.0825062.

In [282]:
#from geopy.geocoders import Nominatim
lat = []
lng = []
#flag = False
geolocator = Nominatim(user_agent="ny_explorer")
location = None

for street,city,zipcode in zip(boston_shopping_mall['address'],boston_shopping_mall['city'],boston_shopping_mall['zipcode']):
    #flag = True
    
    address = street + "," + city + ", MA 0" + str(zipcode)
    #geolocator = Nominatim(user_agent="ny_explorer")
    location = geolocator.geocode(address)
    
    #while(location is None):
    if (location is None):
        print ("None: " + address+ "\n")
    else:
        latitude = location.latitude
        longitude = location.longitude
        #print(latitude)
        #print(longitude)
        #print("\n")
        lat.append(latitude)
        lng.append(longitude)
print(lat)
print(lng)

[42.3471718, 42.3472309, 42.359706, 42.3268965, 42.318836982353865, 42.33872036842105, 42.34859005263158, 42.3479157, 42.354420600000005, 42.35585979166667, 42.355892957694266, 42.36173, 42.361730300000005, 42.3687832, 42.374218, 42.38901760869565]
[-71.08250628733232, -71.07758361909961, -71.0550683, -71.0617906, -71.08457817184478, -71.10737578947369, -71.07430336842106, -71.0843914, -71.0614958, -71.06174775, -71.06016759569124, -71.054485, -71.06660593534323, -71.0759859, -71.120629, -71.11836373913043]


In [283]:
boston_shopping_mall["latitude"] = lat
boston_shopping_mall["longitude"] = lng

In [284]:
boston_shopping_mall.head()

mall name                         address     city  zipcode  \
0  Prudential Center Boston                 800 Boylston St   Boston     2199   
1              Copley Place              100 Huntington Ave   Boston     2116   
2  Faneuil Hall Marketplace                   4 S Market St   Boston     2109   
3          South Bay Center                   8 Allstate Rd   Boston     2118   
4      Washington Park Mall  330 Martin Luther King Jr Blvd   Boston     2119   

   lat  lng   latitude  longitude  
0  NaN  NaN  42.347172 -71.082506  
1  NaN  NaN  42.347231 -71.077584  
2  NaN  NaN  42.359706 -71.055068  
3  NaN  NaN  42.326896 -71.061791  
4  NaN  NaN  42.318837 -71.084578

In [285]:
boston_shopping_mall.drop(columns = ["lat","lng"])

mall name                         address       city  \
0        Prudential Center Boston                 800 Boylston St     Boston   
1                    Copley Place              100 Huntington Ave     Boston   
2        Faneuil Hall Marketplace                   4 S Market St     Boston   
3                South Bay Center                   8 Allstate Rd     Boston   
4            Washington Park Mall  330 Martin Luther King Jr Blvd     Boston   
5               Longwood Galleria               400 Brookline Ave     Boston   
6                 Fresh Pond Mall                140 Clarendon St     Boston   
7      Memorial Convention Center                 900 Boylston St     Boston   
8        Stores At 500 Washington               500 Washington St     Boston   
9               Downtown Crossing                 24 Winter Place     Boston   
10                The Corner Mall               417 Washington St     Boston   
11                Mercantile Mall                      1 Cross St     Boston   
12                   Primax Store                    5 Blossom St     Boston   
13                 Cambridge Side         100 Cambridgeside Place  Cambridge   
14                     The Atrium                    50 Church St  Cambridge   
15  Porter Square Shopping Center                     37 White St  Cambridge   

    zipcode   latitude  longitude  
0      2199  42.347172 -71.082506  
1      2116  42.347231 -71.077584  
2      2109  42.359706 -71.055068  
3      2118  42.326896 -71.061791  
4      2119  42.318837 -71.084578  
5      2215  42.338720 -71.107376  
6      2116  42.348590 -71.074303  
7      2199  42.347916 -71.084391  
8      2111  42.354421 -71.061496  
9      2108  42.355860 -71.061748  
10     2129  42.355893 -71.060168  
11     2109  42.361730 -71.054485  
12     2114  42.361730 -71.066606  
13     2141  42.368783 -71.075986  
14     2138  42.374218 -71.120629  
15     2140  42.389018 -71.118364

### Add shopping malls to Boston map as a feature group layer that can be turn on/off through Layer Control

In [297]:
from folium import FeatureGroup, LayerControl, Map, Marker
map1 = boston_map1

feature_group = FeatureGroup(name='shopping mall')

for lat, lon, mall in zip(boston_shopping_mall['latitude'], boston_shopping_mall['longitude'], boston_shopping_mall['mall name']):
    label = folium.Popup(mall, parse_html=True)
    #folium.CircleMarker
    Marker(location=[lat,lon],popup=label).add_to(feature_group)
    
feature_group.add_to(map1)
LayerControl().add_to(map1)
       
map1

# 6. Conclusion

### I read the map carefully after shooping malls are added on it, and find there is a shopping mall called Washington Park Mall near the candidate location 1. On the other hands, there is no a big shopping mall near the candidate location 3, which causes a concern of less customer traffic. Now I can draw a conclusion that the candidate location 1 is the best location for the new Italian restaurant.